File: frontend/portal/src/bewertungFrist.ts
import type { Bericht } from './api';
import { UEMS_BEWERTUNG_SAETZE } from './glossar';

/**
 * UEMS AP-16 IP-24 (S5/S6, §5.5 Schritt 4): der Übersichts-Baustein „Energetische Bewertung“ am Unternehmen — rein.
 * Frist, Fälligkeit, Zahlen und Verantwortliche kommen fertig aus `GET /api/v1/berichte` (`ueberpruefung`, beim Abruf
 * abgeleitet); hier wird nur die gültige Bewertung gewählt und in Sätze aus `glossar.ts` gesetzt — nichts gerechnet.
 */

export const BEWERTUNG_VORLAGE = 'energetische_bewertung';

export interface BewertungFristBild {
  kennung: string;
  /** „Energetische Bewertung: Stand Nr. 2 vom 17.11.2026 · Überprüfung fällig seit 1 Tag.“ */
  satz: string;
  faellig: boolean;
  /** „6 wesentliche Energieeinsätze · 1 offener Messbedarf“ */
  zahlen: string;
  /** S6 — nur wenn fällig: die Verantwortlichen der wesentlichen Einsätze. */
  hinweis: string | null;
  ohneVerantwortliche: string | null;
}

/** `2026-11-17` → `17.11.2026` (ein Kalendertag, keine Zeitzone). */
export const tagText = (iso: string): string => `${iso.slice(8, 10)}.${iso.slice(5, 7)}.${iso.slice(0, 4)}`;

/**
 * Die gültige Bewertung: Vorlage `energetische_bewertung`, mit freigegebenem Stand, nicht abgelöst (der Server kennzeichnet
 * jede ältere mit `abgeloest_durch`). Ohne sie gibt es keinen Baustein (`null`) — Bestandskunden merken nichts (R11).
 */
export function bewertungFristBaustein(berichte: readonly Bericht[] | null): BewertungFristBild | null {
  const b = (berichte ?? []).find(
    (x) => x.vorlage === BEWERTUNG_VORLAGE && x.ueberpruefung != null && x.ueberpruefung.abgeloest_durch == null,
  );
  const u = b?.ueberpruefung;
  if (!b || !u) return null;
  const vom = tagText(u.stand_vom);
  const satz = !u.ueberpruefung_faellig
    ? UEMS_BEWERTUNG_SAETZE.fristAm(u.stand_nr, vom, tagText(u.faellig_am ?? u.stand_vom))
    : u.faellig_seit_tagen === 0
      ? UEMS_BEWERTUNG_SAETZE.fristHeute(u.stand_nr, vom)
      : UEMS_BEWERTUNG_SAETZE.frist(u.stand_nr, vom, u.faellig_seit_tagen ?? 0);
  const namen = u.verantwortliche.map((v) => `${v.name} (${v.einsaetze.join(', ')})`).join(' · ');
  return {
    kennung: b.kennung,
    satz,
    faellig: u.ueberpruefung_faellig,
    zahlen: UEMS_BEWERTUNG_SAETZE.fristZahlen(u.wesentliche_einsaetze, u.offene_bedarfe),
    hinweis: u.ueberpruefung_faellig && namen ? UEMS_BEWERTUNG_SAETZE.fristHinweis(namen) : null,
    ohneVerantwortliche:
      u.ueberpruefung_faellig && u.ohne_verantwortliche.length > 0
        ? UEMS_BEWERTUNG_SAETZE.fristOhneVerantwortliche(u.ohne_verantwortliche.join(', '))
        : null,
  };
}

File: services/api/src/main/java/com/voltpilot/api/uems/BewertungFrist.java
package com.voltpilot.api.uems;

import java.time.Instant;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * AP-16 S5/S6 (IP-24): die Frist der energetischen Bewertung — beim Abruf abgeleitet, nie geschrieben (kein Läufer, kein
 * Ereignis). Frist = Tag der Freigabe des jüngsten gültigen Stands (Zone des Berichts) + {@code wiedervorlage_monate}; ab
 * diesem Tag ist die Überprüfung fällig (am Tag selbst „seit 0 Tagen“). Die Wiedervorlage wirkt beim nächsten Abruf — sie
 * ist eine Eigenschaft der Bewertung, kein Teil des eingefrorenen Stands.
 *
 * <p>Gibt ein Stand einer anderen Bewertung desselben Unternehmens später frei, ist diese Bewertung abgelöst (R10 Schritt 3):
 * keine Frist mehr, lesbar bleibt sie. Die Uhr kommt immer von außen ({@code heute}); hier liest nichts die Rechneruhr.
 */
public final class BewertungFrist {

    private BewertungFrist() {}

    /** Der Stand, an dem die Frist hängt, und — nur bei einer abgelösten Bewertung — ihr Nachfolger. */
    public record Frist(int standNr, LocalDate standVom, int wiedervorlageMonate, LocalDate faelligAm, boolean faellig,
            Integer faelligSeitTagen, String abgeloestDurch) {}

    /** Der jüngste Stand einer Bewertung — Grundlage der Ablösung. */
    public record BewertungStand(String kennung, int nr, Instant freigegebenAm) {}

    /** Die Frist am Tag {@code heute}; {@code abgeloestDurch} ≠ {@code null} = keine Frist, nichts fällig. */
    public static Frist ableiten(int standNr, LocalDate standVom, int wiedervorlageMonate, LocalDate heute,
            String abgeloestDurch) {
        if (abgeloestDurch != null) {
            return new Frist(standNr, standVom, wiedervorlageMonate, null, false, null, abgeloestDurch);
        }
        LocalDate faelligAm = standVom.plusMonths(wiedervorlageMonate);
        boolean faellig = !heute.isBefore(faelligAm);
        return new Frist(standNr, standVom, wiedervorlageMonate, faelligAm, faellig,
                faellig ? Math.toIntExact(ChronoUnit.DAYS.between(faelligAm, heute)) : null, null);
    }

    /** Eine verantwortliche Person und ihre wesentlichen Einsätze (S6). */
    public record Verantwortung(String name, List<String> einsaetze) {}

    /**
     * S6: die Verantwortlichen der WESENTLICHEN Einsätze, je Person einmal, in der Reihenfolge ihres ersten Einsatzes. Ein
     * wesentlicher Einsatz ohne verantwortliche Person steht nicht hier, sondern in {@link #ohneVerantwortliche}.
     */
    public static List<Verantwortung> verantwortliche(List<BerichtRepository.EinsatzLage> lage) {
        Map<String, List<String>> je = new LinkedHashMap<>();
        for (BerichtRepository.EinsatzLage e : lage) {
            if (e.wesentlich() && e.verantwortlichName() != null) {
                je.computeIfAbsent(e.verantwortlichName(), n -> new ArrayList<>()).add(e.kennzeichen());
            }
        }
        return je.entrySet().stream().map(x -> new Verantwortung(x.getKey(), List.copyOf(x.getValue()))).toList();
    }

    public static List<String> ohneVerantwortliche(List<BerichtRepository.EinsatzLage> lage) {
        return lage.stream().filter(e -> e.wesentlich() && e.verantwortlichName() == null)
                .map(BerichtRepository.EinsatzLage::kennzeichen).toList();
    }

    /**
     * Die Bewertung, die {@code kennung} ablöst: die mit dem spätesten jüngsten Stand unter allen nicht archivierten
     * Bewertungen desselben Unternehmens — {@code null}, wenn {@code kennung} selbst diese ist. Gleichstand: die höhere
     * Kennung gilt (eine feste Reihenfolge, nie zwei gültige).
     */
    public static String abgeloestDurch(String kennung, List<BewertungStand> staende) {
        BewertungStand gueltig = null;
        for (BewertungStand s : staende) {
            if (gueltig == null || s.freigegebenAm().isAfter(gueltig.freigegebenAm())
                    || (s.freigegebenAm().equals(gueltig.freigegebenAm()) && s.kennung().compareTo(gueltig.kennung()) > 0)) {
                gueltig = s;
            }
        }
        return gueltig == null || gueltig.kennung().equals(kennung) ? null : gueltig.kennung();
    }
}
